[20.03] haskellPackages.ghcide: Fix build

maralorn opened the pull request:
###### Motivation for this change

`ghcide` is marked as broken and is not building. With these changes it builds.

Please let me know if I should structure the overrides in a different way.
When this PR is approved I will port it to the haskell-updates branch. But first I wanna know if I did this right.
I am especially confused by the recursive library overrides. It‘s not pretty this way. e.g. the override for regex-tdfa_1_3_1_0 which needs a pinned regex-base version is now twice in the file.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

maralorn committed:
haskellPackages.ghcide: Fix ghcide and hie-bios

cdepillabout commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -1440,6 +1440,27 @@ self: super: {
+  # this will probably need to get updated with every ghcide update,
+  # we need an override because ghcide is tracking haskell-lsp closely.
+  ghcide = dontCheck (super.ghcide.override rec {
+    haskell-lsp-types = super.callHackageDirect {
+      pkg = "haskell-lsp-types";
+      ver = "0.19.0.0";
+      sha256 = "1z0c9c2zjb4ad3ffzng9njyn9ci874xd8mmqwnvnm3hyncf1430g";
+    } {};
+    haskell-lsp = super.callHackageDirect {
+      pkg = "haskell-lsp";
+      ver = "0.19.0.0";
+      sha256 = "1v0r57g2dhradnjnvp40jmv5swawg9k3d735kj50dca1gbx66y0c";
+    } {
+      inherit haskell-lsp-types;
+    };
```
I don't think `callHackageDirect` will work in nixpkgs because it uses IFD.

Alternatively, you can use `cabal2nix` to create a derivation for `haskell-lsp-types_0_19_0_0` and `haskell-lsp_0_19_0_0` by hand and throw it in `pkgs/development/haskell-modules/hackage-packages.nix`.

The `regex-tdfa` and `haddock-library` overrides look fine.

maralorn replied:
Oh, I didn‘t realize the IFD problem. Okay.
But isn‘t `hackage-packages.nix` autogenerated? Won‘t it get overriden? Is that really the right place to put it?

maralorn replied:
Mind you. Probably it won‘t get overwritten on stable anymore. But I also need a solution that works on the haskell-updates branch.

maralorn replied:
Thanks for the feedback. I guess I figured out that I also need to add it to extra-packages in the hackage2nix.yaml so that it does not get overriden.

cdepillabout replied:
Unfortunately there isn't a single solution that works for both the release branches and `haskell-updates`.

On the release branches, `hackage-packages.nix` is not autogenerated anymore.  It should be okay to edit it directly.

On `haskell-updates`, the way to get new packages generated is to add the package to this list:

https://github.com/NixOS/nixpkgs/blob/0a115f808755ad6eec5eb7742c71d02e4d0f9d9f/pkgs/development/haskell-modules/configuration-hackage2nix.yaml#L2361-L2370

I have somewhat of an explanation of this process here:

https://discourse.nixos.org/t/haskellpackages-stm-containers-fails-to-build/5416/4

maralorn replied:
Okay. Well, now I tried to do a change that fits for both branches. i.e. change nix file and the yaml file both. I can amend my commits so that I do the one for the release and the other for haskell-updates.
But not change the .nix file for haskell-updates is annoying because then I can‘t even test if it‘s working right now. Or is it desired for me to run hackage2nix after changing the yaml?

cdepillabout replied:
> But not change the .nix file for haskell-updates is annoying because then I can‘t even test if it‘s working right now. Or is it desired for me to run hackage2nix after changing the yaml?

I don't have a good suggestion for you.

However, there are people that are working on a way to make running `hackage2nix` easier:

https://github.com/NixOS/nixpkgs/pull/86699

-------------------------------------

One easy(?) thing you could do is send a PR adding the packages you need to `extra-packages`, wait for it to be merged in and hackage-packages.nix regenerated, and then send a second PR using those newly regenerated packages.

cdepillabout commented on pkgs/development/haskell-modules/hackage-packages.nix:
```diff
@@ -96014,7 +96014,6 @@ self: {
        description = "The core of an IDE";
        license = stdenv.lib.licenses.asl20;
        hydraPlatforms = stdenv.lib.platforms.none;
-       broken = true;
```
You'll have to remove the `hydraPlatforms = stdenv.lib.platforms.none;` lines here as well.

If you don't remove them, then Hydra won't build these packages on any platforms, and you won't get builds from the cache.

maralorn replied:
Oh, that makes total sense.

cdepillabout commented:
@maralorn Thanks for working on this.  I think a lot of Haskellers have been trying out ghcide.

maralorn commented:
@cdepillabout I do believe I have now fixed all your concerns and all my concerns.

* Moved the lsp stuff to hackage-packages and put that in the yaml, so that it doesn‘t get lost.
* Activated hydra builds.
* figured out the use of self, so that I need to do less overrides on my own.

cdepillabout commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -1440,6 +1440,21 @@ self: super: {
 
+  # Needed for ghcide
+  haskell-lsp_0_19_0_0 = super.haskell-lsp_0_19_0_0.override {
+    haskell-lsp-types = super.haskell-lsp-types_0_19_0_0;
+  };
+
+  # this will probably need to get updated with every ghcide update,
+  # we need an override because ghcide is tracking haskell-lsp closely.
+  ghcide = dontCheck (super.ghcide.override rec {
+    haskell-lsp-types = super.haskell-lsp-types_0_19_0_0;
```
```suggestion
    haskell-lsp-types = self.haskell-lsp-types_0_19_0_0;
  };

  # this will probably need to get updated with every ghcide update,
  # we need an override because ghcide is tracking haskell-lsp closely.
  ghcide = dontCheck (super.ghcide.override rec {
    haskell-lsp-types = self.haskell-lsp-types_0_19_0_0;
```

In general, derivations that are being overriden need to come from `super`, and dependencies need to come from `self`.

maralorn replied:
Since there are no overrides for lsp-types in the file, I didn‘t  think it was necessary. But yes, it makes sense, when the package breaks and needs to overriden manually.

cdepillabout commented:
One more change and this can be merged in:

`haddock-library_1_8_0` needs to have the `hydraPlatforms = stdenv.lib.platforms.none` line removed.

Once that is done, I will merge this in.  I have confirmed that `ghcide` was able to build.

cdepillabout commented on pkgs/development/haskell-modules/hackage-packages.nix:
```diff
@@ -107671,7 +107669,6 @@ self: {
        testToolDepends = [ hspec-discover ];
        description = "Library exposing some functionality of Haddock";
        license = stdenv.lib.licenses.bsd2;
-       hydraPlatforms = stdenv.lib.platforms.none;
```
Looks good, thanks.
